CHANGELOG.md

```diff
@@ -1,12 +1,19 @@
 # Changelog
 
-## 24.5.3 [#734](https://github.com/openfisca/openfisca-core/pull/734)
+### 24.5.3 [#737](https://github.com/openfisca/openfisca-core/pull/737)
+
+- Fix an encoding bug of variables with value type str
```

Comment:
Agreed, the fix went beyond what was strictly necessary.

I'm not sure the "thoughts and time" remark is quite justified though, we have systemic contributors to the incident (lack of unit tests or documentation covering this behaviour, Python 2/3 continuing hassles), and both @maukoquiroga and I paired with @sandcha independently so I don't see this as a lack of individual attention or rushing.

I'm not sure I understand your remark on Unicode, the Python 3 default encoding is UTF8 (and this carries over to Numpy via the "U" dtypes according to [the docs](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.dtypes.html#string-dtype-note)) and while I admit to low expertise in string encodings my understanding is that this "variable-width" encoding does not necessarily use four bytes per character.

Reply:
Hi @fpagnoux.

> Use unicode for all string variables

That's definitely better, although I didn't know "string" was a Python concept so that's why I didn't use it here.

> Deprecate max_length variable attribute

Didn't know we were deprecating it, nor that there was a truncation feature.

> Which I think would have deserve maybe a little more thoughts and time.

We were three to work on this specific issue, so more brains weren't available, and invested time was consequent.

Doing the math, we invested ballpark 4h * 3 = 12h of 75% of the team's brains, which is basically around ~10% of our sprint allocated time. I have to agree with @Morendil here:

> we have systemic contributors to the incident (lack of unit tests or documentation covering this behaviour, Python 2/3 continuing hassles)

However, I assume the responsibility as I reviewed, cleaned up, rebased, renamed,  fixed up and merged.

That being said, I don't know exactly what could I have done to avoid the situation.

Reply:
What I meant by time was more _absolute time_ than _brain time_:

- A review is requested at D0, **2 days ago**. A comment (where I'm mentioned) clearly seem to indicate that the PR openers don't fully understand the purpose of the code they are removing.
- At D0+1, the PR is merged, with this comment still unanswered. 
- Later that day, I check the PR, following the mention. I'm surprised by the wide scope of the PR, and then realize France broke. 
- At D0+2, we revert everything

The pity is that we were very close from avoiding the incident. Delaying the merge a day, until @sandcha's comment was answered, would have been enough😬 .

Having said that, misevaluating the scope of a PR happens. Now we can just:
- extend unit tests to making it less likely
- keep in mind than reverting/unpublishing is the best option it done quickly when it happens 🙂

Reply:
> I'm not sure I understand your remark on Unicode, the Python 3 default encoding is UTF8 (and this carries over to Numpy via the "U" dtypes according to the docs) and while I admit to low expertise in string encodings my understanding is that this "variable-width" encoding does not necessarily use four bytes per character.

To illustrate my remark:

```py
import numpy as np

postal_codes = ['75012', '75015', '75007']
bytes_array = np.asarray(postal_codes, dtype=np.bytes_)
unicode_array = np.asarray(postal_codes, dtype=np.unicode_)

print(unicode_array.nbytes)
>>> 60  # 3 items x  20 bytes

print(bytes_array.nbytes)
>>> 15 # 3 items x 5 bytes
```

Reply:
We should probably open a separate issues to discuss all the issues that we have identified.
- Should all string variables be unicode, for convenience?
- Should all string variables be bytes, for memory optimisation?
- Should we allow both? Sounds nice, but is there really a demand for unicode variables?
- Is the "max_length" field really useful? Should it throw errors instead of silently shortening user inputs?

Reply:
Ah, the plot thickens. This is apparently not documented but [here's an interesting discussion](https://github.com/PyTables/PyTables/issues/499) pointing out that Numpy uses UTF-32 internally, rather than UTF-8 as I'd inferred from the Numpy docs.

The thread points to an additional option:
```
>>> object_array = np.asarray(postal_codes, dtype="O")
>>> object_array.nbytes
24
>>> type(object_array[0])
<class 'str'>
```

Still some overhead, but limited.

> Should all string variables be unicode, for convenience?

The alternative is that as soon as someone tries to use `value_type = str` for something - anything - that can conceivably contain even one accented character, they're getting set up for a [nasty surprise](https://github.com/betagouv/beta.gouv.fr/pull/1688). I think that goes beyond convenience, it's an egregious violation of the principle of least surprise.

If it's important for performance reasons, then one solution is to encode strings internally as byte arrays before storing them into Numpy arrays and decode them back into proper strings whenever their values is needed. To do that, we'd need to hide Numpy arrays (an implementation detail) behind the Variable/Holder (I'm not sure which) interface protocol. This is possibly a good idea (on information hiding principles) irrespective of the current issue, even if it would admittedly involve a lot of busywork.

Reply:
> Delaying the merge a day, until @sandcha's comment was answered, would have been enough 😬.

Can't disagree, but we are analysing this now with the benefit of hindsight. In [Allspaw's words](https://codeascraft.com/2012/05/22/blameless-postmortems/) "the action made sense to the person at the time they took it, because if it hadn’t made sense to them at the time, **they wouldn’t have taken the action in the first place**".

There are any number of additional counterfactual scenarios where merging the PR wouldn't have resulted in a meltdown, including "if we'd had more tests on string-valued variables" and "if we'd limited the fix to the least necessary change" or "if the people pairing on the fix had been more sensitive about memory usage".